chanzuckerberg/single-cell-data-portal
# chore(fe): upgrade to nextjs12 blueprint4 node16

- #1705

### Reviewers
**Functional:** 
@seve @hthomas-czi @frano-m 

**Readability:** 
@atarashansky 

---

## Changes
1. Upgrade Next.js from `11` to `12`
2. Upgrade Blueprint.js from `3` to `4`. [Changelog](https://github.com/palantir/blueprint/wiki/Blueprint-4.0)
3. Pin node from `14` to `16`. UPDATE: Pinning to `16.14.2` to avoid GH Action lint error exit code 243. [[Source]](https://stackoverflow.com/a/71892226/3120863)
4. Remove `styled-components` and use `emotion` completely (so we can use Next's fast SWC compiler and only have one css-in-js dependency)

Notable changes:
1. Next12 doesn't use Babel anymore and defaults to use its new SWC rust based compiler, which is supposedly 17x faster [migration doc](https://nextjs.org/docs/upgrading#upgrading-from-11-to-12). So in order for `jest` test runner to work with TS files, we need to update jest's config files to use `next/jest` to compile TS files before jest consumes them

3. Blueprint3 to Blueprint4 upgrade was a pain. It includes removing `node-sass` as a direct dependency (which is now officially deprecated in favor of `sass`(dart-sass)) AND also install `@vgrid/sass-inline-svg` to process Blueprint SVG files. On top of that, because we define a custom theme via sass, I had to follow [custom sass processing setup steps](https://github.com/palantir/blueprint/issues/4759#issuecomment-1112218581) to overcome Blueprint's very limited support of theming. Currently the Blueprint team hasn't come up with a solution for official theming support, as shown in [this comment](https://github.com/palantir/blueprint/issues/4032#issuecomment-1087723314). Given that `@vgrid/sass-inline-svg` actually still depends on `node-sass`, we might need to move away from Blueprint completely sooner than later to avoid indirectly depending on a deprecated package for too long

3. Migration from `styled-components` to `emotion` includes a small change of ~removing the usage of `import { css } from 'styled-components'`, because `@emotion/css` outputs a class string instead of a styling function. So I just rewrote those instances with a styling function instead! See `frontend/src/components/Collections/components/Dataset/common/style.ts` for example~

    UPDATE: Thanks to Alec, we just had to change the import from `{ css } from 'styled'` to `{ css } from '@emotion/react'`

## QA steps (optional)
RDEV:
https://thuang-next-12-frontend.rdev.single-cell.czi.technology/

Local:
1. Make sure your node is v16
2. Run `rm -rf node_modules && npm i` to install all new dependencies
3. Run the app via `make local-sync` OR if your local app connects to dev/staging BE, then just `npm run dev`
5. App should work and all the styles should look the same!

## Notes for Reviewer
Tagging all the people that work on FE for visibility!
